### PR TITLE
perf(wand): use block-level tf bounds during BMW advancement (BUG-005)

### LINKED ISSUE #143: BUG-005: WAND skip_to_block ignores block-level max-tf upper bounds
# BUG-005: WAND skip_to_block ignores block-level max-tf upper bounds

**Category:** Performance
**Severity:** Medium
**File(s):** searchlite-core/src/query/wand.rs
**Line(s):** 250-258, 829-844

## Description
`TermState::skip_to_block` only uses `block_meta.doc_ids` (per-block maximum `doc_id`) to advance the cursor — it does not consult `block_meta.tfs` (per-block maximum `term_freq`). During the pivot-seeking phase of `wand_loop`, for terms that appear strictly before the pivot (`queue[0..p_idx]`), the code calls:
```rust
let moved = term.skip_to_block(pivot_doc);
```
This only skips forward to the block that *contains* `pivot_doc` by doc_id. But a BMW-style optimisation can be more aggressive: any block whose max-tf contribution (via `score_tf(tf_max, ..., min_doc_len, ...)`) plus the accumulator cannot reach the current `heap_threshold` can be skipped entirely, even if its doc_ids overlap the target range.

Because `use_block_bounds` is *only* enabled for `ExecutionStrategy::Bmw` (wand.rs:443), users who explicitly opt into BMW pay the cost of building block metadata (see `build_block_meta`, wand.rs:298-329) but never reap the bound-based skip benefit during advancement.

## Reproduction / Evidence
```rust
// wand.rs:250-258
fn skip_to_block(&mut self, target: DocId) -> usize {
  let prev = self.idx;
  let block_idx = self.block_meta.doc_ids.partition_point(|doc| *doc < target);
  let start = block_idx.saturating_mul(self.block_meta.block_size);
  if start > self.idx {
    self.idx = start.min(self.postings.len());
  }
  self.idx.saturating_sub(prev)
}
```
No use of `self.block_meta.tfs`. The complementary `block_upper_bound()` method (wand.rs:231-244) computes per-block score ceilings but is *only* consulted during pivot selection (wand.rs:722-735), never during advancement.

Benchmark evidence: for queries with high-cardinality terms (e.g., `"the OR cat"` with one common and one rare term, >1M postings each), BMW's observed speedup over plain WAND is <5% in `searchlite-core/benches/end_to_end.rs`, where published BMW implementations typically show 2-4× improvement.

## Expected Behaviour
During advancement, skip blocks whose `score_tf(block_tf_max, df, min_doc_len, ...)` contribution cannot push the running accumulator past the current top-k threshold.

## Actual Behaviour
Only doc_id-based block skipping is performed; tf-bound skipping is effectively dead.

## Suggested Fix
Add a bound-aware skip entry point consumed from `wand_loop` when `use_block_bounds && rank_hits && heap.len() >= k`:
```rust
fn skip_blocks_below_bound(&mut self, min_contribution: f32) -> usize {
  let prev = self.idx;
  let mut block_idx = self.block_index();
  while block_idx < self.block_meta.tfs.len() {
    let tf = self.block_meta.tfs[block_idx];
    let bound = score_tf(tf, self.df, self.min_doc_len, self.avgdl,
                         self.docs, self.k1, self.b, self.weight);
    if bound >= min_contribution { break; }
    block_idx += 1;
    self.idx = (block_idx * self.block_meta.block_size).min(self.postings.len());
  }
  self.idx.saturating_sub(prev)
}
```
Then at wand.rs:831-841 call `skip_blocks_below_bound(heap_threshold - acc_so_far)` before `advance_to(pivot_doc)`.

## Related
Migrated from `.old-bugs/wand_block_optimization_missing.md` (same finding; re-filed here to bring it under the current `bugs/` index).

### PR DESCRIPTION
**Pull request by davidkelley**
## Summary

Fixes BUG-005 (#143). `TermState::skip_to_block` only used `block_meta.doc_ids` (per-block maximum doc_id) to advance the cursor during the WAND advancement phase. `block_meta.tfs` (per-block maximum term frequency) was computed and stored by `build_block_meta` but only consulted during pivot selection via `block_upper_bound()`, never during advancement. Users who explicitly opted into the `Bmw` execution strategy paid the cost of building block metadata without reaping the bound-based skip benefit during the advancement phase.

## What changed

- **`searchlite-core/src/query/wand.rs`** — add `TermState::skip_blocks_below_bound(min_contribution)`, which skips forward through blocks whose per-block max-tf score contribution (via `score_tf`) is below `min_contribution`. Stops at the first block whose upper bound meets the threshold, or at the end of the postings list.

- **`searchlite-core/src/query/wand.rs`** (wand_loop) — integrate tf-based block skipping into the advancement phase. When `use_block_bounds` is active (BMW strategy), the top-k heap is full, and `rank_hits` is true:
  1. Compute the total global upper bound of all queued terms once.
  2. For each term being advanced to `pivot_doc`, derive `min_needed = max(0, heap_threshold - sum_of_other_global_upper_bounds)`.
  3. Call `skip_blocks_below_bound(min_needed)` before the existing `skip_to_block(pivot_doc)` and `advance_to(pivot_doc)`.

  This ensures blocks whose max-tf contribution cannot push a candidate past the heap threshold are skipped entirely, reducing the number of postings examined during advancement.

- **`searchlite-core/src/query/wand.rs`** (tests) — four new regression tests:
  - `bmw_produces_same_top_k_as_wand` — correctness: BMW and WAND return identical top-k results and scores.
  - `bmw_block_tf_skipping_reduces_postings_advanced` — performance: BMW advances no more postings than plain WAND on a workload with mixed high/low tf blocks, verifying the skip path is exercised.
  - `skip_blocks_below_bound_skips_low_tf_blocks` — unit: directly tests the new method with known block-level tf values, asserting blocks 0-1 (tf=1) are skipped and block 2 (tf=10) is the landing point.
  - `skip_blocks_below_bound_noop_when_threshold_zero` — edge case: threshold <= 0 returns immediately with no advancement.

## Test plan

- [x] `cargo test -p searchlite-core --lib query::wand` — 8/8 pass (4 existing + 4 new).
- [x] `cargo test -p searchlite-core --lib` — 291/291 pass.
- [x] `cargo test -p searchlite-core` — full core test matrix green (lib + all integration suites).
- [x] `cargo test -p searchlite-ffi -p searchlite-cli -p searchlite-http` — all green.
- [x] `cargo fmt --all --check` clean.
- [x] `cargo clippy -p searchlite-core --tests --no-deps -- -D warnings` clean.

Closes #143

### CODE REVIEW

File: searchlite-core/src/query/wand.rs [hunk 1]
```diff
@@ -996,4 +1057,327 @@ mod tests {
...
+    // Block 3 (doc_ids 12-15): tf=10 each (high contribution)
+    let mut common_entries = Vec::new();
+    for doc_id in 0..12 {
+      common_entries.push(PostingEntry {
+        doc_id,
+        term_freq: 1,
+        positions: smallvec![],
+      });
+    }
+    for doc_id in 12..16 {
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/src/query/wand.rs]: Good catch — the original test never entered the `pivot_doc > smallest_doc` branch because `common` always held the smallest `doc_id`.

Restructured the test (now `bmw_block_tf_skipping_exercises_advancement_path`) with an anchor/spread layout:
- Both terms overlap at doc_ids 0-3 with tf=20, filling the heap (k=3) with high combined scores
- Anchor then jumps to doc_id 40; spread is at doc_id 4
- Queue becomes `[spread(4), anchor(40)]` → `pivot_doc(40) > smallest_doc(4)` → advancement path entered
- Spread must advance through 9 blocks of tf=1 (doc_ids 4-39) — these get skipped by `skip_blocks_below_bound`

The assertion is now strict `<` (not `<=`), confirming blocks are actually skipped. See 24abae9.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by davidkelley** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_
